[New Feature] Custom Server Action: make customizing on server side on SaaS possible

szufisher opened the pull request:
![image](https://user-images.githubusercontent.com/12823863/35571096-1f4ebee6-060c-11e8-986d-0a3dfba3055a.png)
**Features added**
1. Create new/update existing document or execute python code when document created/saved/submitted/value changed, or specified method called
or scheduled time event:before or after certain days based on specified date field reached,
2. Execute python code per javascript calling convention

**Use Cases:**
1. Auto assign responsible person for new or updated task/order/issue etc
2. Auto create error log/to do /issue when certain documents created/changed under specific conditions
3. do customized validation check for custom doctypes: using frappe.throw(_('validation failure!'))
4. do customized handling:auto filling certain fields for custom doctypes
5. do customized complex server side logic for custom button: get items from , create subsequent documents, jump to related document etc
6. do required simple or complex customization on SaaS, which currently not possible! 
7. do what you want per your creative idea!

**How to use it**
1. define custom server action
	Action Type: with following options
		Update Record: when event triggered, the doc(target document type per the link field, the current doc as default) will be updated,
						to be updated fields are defined under the field mapping child table
		Create New Record:when event triggered, a new doc of target document type will be created, fields are defined under the field mapping child table
		Execute Python Code: when even triggered , the Python code will be called
		Python Code Called By JS: the python code will be called via frappe.call from javascript, normally bound to custom button in form view.
	Document Type: the document type by which to trigger the action,
	Trigger On: when to trigger, with following options
 		New
 		Save
 		Submit
 		Cancel
 		Days After
 		Days Before
 		Value Change
 		Method
	        Custom

	Trigger Method: when selected trigger on as Method, specify the method name here
	Value Changed: when selected trigger on as Value change, specify the field (value change)
	Reference Date : when selected trigger on as Days Before or Days After, specify the referenced date field
	Days Before Or After: when selected trigger on as Days Before or Days After, specify the number of days
	Condition: python expression, only the evaluated result is True, the action will be called
	Code: the python code to be exectuted, when action type is Execute Python Code or Python Code Called By Client
	Target Document Type: when action type is Update Record or Create New Record, default as the document type
	Link Field: when action type is Update Record and target document type is not same as document type, the link field from the document type
	Field Mapping
		Field: the target field of the target document type
		Type: Value, Formula
		Value: if type is value, the field value, if type is formula, the python expression, just like the condition field
		
    tips for python code and python expression
		local context variable: doc, frappe, json,_ is available
		internal function is not supported, thus def, return is not allowed, for passing input parameter and return result, use frappe.local.form_dict instead
		special key custom_server_action_out is used for returning value from python code, for more details with samples refer to the relevant post in discuss.erpnext.com
		
How it works in the background technically
1. Added 2 new doctypes, referred to exsting email alert feature, instead of updating one field with fixed value only,added a child table to allow assign
multi fields for new / update document, enhanced field value assignment to support python expression, also added the possibility to execute python code for
more complicated cases
	1.1 custom server action
	1.2 custom server action field
2. added one new version of the safe_eval (based on Odoo), whitelisted opcodes are defined, thus to safeguard the security of the customized python code
3. Updated model.document.py,  added hooks for the custom server action , like existing web hooks.
4. added 2 handlers in frappe/custom/doctype/custom_server_aciton.__init__ run_custom_server_action and run_custom_server_action_by_js.

Commit:
Add files via upload

Commit:
Create custom_server_action

Commit:
Delete custom_server_action

Commit:
Create init

Commit:
Add files via upload

Commit:
Delete custom_server_action.js

Commit:
Delete custom_server_action.json

Commit:
Delete custom_server_action.py

Commit:
Delete test_custom_server_action.js

Commit:
Delete test_custom_server_action.py

Commit:
Create init

Commit:
Add files via upload

Commit:
Rename init to __init__.py

Commit:
Rename init to __init__.py

Commit:
safe_eval

Commit:
handler run_custom_server_action

Commit:
document.py add run_custom_server_action

Commit:
add support of translation _

Commit:
Update custom_server_action.js

Commit:
Update custom_server_action.py

Commit:
Update handler.py

Commit:
Update document.py

Commit:
Update safe_eval.py

rmehta commented on frappe/custom/doctype/custom_server_action/custom_server_action.js:
```diff
@@ -0,0 +1,122 @@
+		frm.trigger('event');
+	},
+	document_type: function(frm) {
+	if(!frm.doc.target_document_type) {
```
whitespace

szufisher replied:
done

rmehta commented on frappe/custom/doctype/custom_server_action/custom_server_action.js:
```diff
@@ -0,0 +1,122 @@
+
+	target_document_type: function(frm){
+	    if(!frm.doc.target_document_type) {
+			return;
```
whitespace

szufisher replied:
done

rmehta commented on frappe/custom/doctype/custom_server_action/custom_server_action.py:
```diff
@@ -0,0 +1,192 @@
+		frappe.throw(_("Error in Custom Server Action"))
+
+def send(server_action, doc):
+		'''Build recipients and send Custom Server Action'''
```
whitespace (why double tabs)?

rmehta commented on frappe/handler.py:
```diff
@@ -18,7 +19,16 @@ def handle():
+	if frappe.local.form_dict.get('method','').split('.')[0]=='custom_server_action':
+		server_action =  frappe.local.form_dict.get('method','').split('.',1)[-1]
+	if server_action:
+		data = run_custom_server_action(server_action)
```
the frappe-way of writing this is by adding `@whitelist` method

example

Example in `frappe/custom/doctype/custom_server_action/custom_server_action.py`

```
@frappe.whitelist()
def run_custom_server_action(name)
```

then call with

`/api/method/frappe.custom.doctype.custom_server_action.custom_server_action.run_custom_server_action`

This kind of hard-coding in handler.py is not good!

szufisher replied:
moved to custom/doctype/custom_server_action/__init__.py

rmehta commented on frappe/handler.py:
```diff
@@ -31,6 +41,29 @@ def handle():
 
 	return build_response("json")
 
+def run_custom_server_action(server_action_name):
```
move this out of handler.py

rmehta commented on frappe/model/document.py:
```diff
@@ -808,6 +811,52 @@ def _evaluate_alert(alert):
 			elif alert.event=='Method' and method == alert.method:
 				_evaluate_alert(alert)
 
+	def run_custom_server_action(self, method):
```
move this code out of `document.py` it is very long

szufisher replied:
moved to custom/doctype/custom_server_action/init.py

rmehta reviewed (changes requested):
Needs test cases! lots of them. We need to ensure malicious code is not executed with custom server action.

Also update docs

Commit:
Update custom_server_action.py

Commit:
move the run custom server action to custom module

Commit:
move run custom server action to custom module

Commit:
Update __init__.py

Commit:
Update custom_server_action.py

Commit:
Update __init__.py

Commit:
Update mapper.py

Commit:
Update document.py

Commit:
Update __init__.py

Commit:
Update custom_server_action.py

Commit:
Update handler.py

Commit:
Update __init__.py

keep custom server action in cache, 
throw via msgprint instead of frappe.throw because code executed via eval is at global module level , the exception raised in which can not be catched/processed by the caller( app.handle)

Commit:
Update custom_server_action.json

Commit:
Update custom_server_action.js

Commit:
Update custom_server_action.py

Commit:
Update __init__.py

Commit:
Update __init__.py

Commit:
Update custom_server_action.json

Commit:
Update custom_server_action.py

Commit:
Update custom_server_action.py

Commit:
Update custom_server_action.js

Commit:
Update __init__.py

Commit:
Update custom_server_action.py

Commit:
Update handler.py

Commit:
Update document.py

Commit:
Update mapper.py

Commit:
Update __init__.py

Commit:
Update document.py

Commit:
Update custom_server_action.py

Commit:
Update custom_server_action.py

szufisher commented:
Here the documentation for how to make Javascript callable custom server actions
1. create new custom server action, select action type as Python Code Called By JS
    Put the python code into the code field,  context variables: frappe, _, json is available,  for the input 
     parameter and return result , frappe.local.form_dict should be used, like below
            purchase_order = frappe.local.form_dict.get('purchase_order')
            item_code = frappe.local.form_dict.get('item_code')

            frappe.local.form_dict.update({'custom_server_action_out':stock_entry.as_dict()})
2. in javascript,in the function to be triggered by custom button,  call the method run_custom_server_action_by_js together with the server_action_name argument like below: 
              method: "frappe.custom.doctype.custom_server_action.run_custom_server_action_by_js",
	      server_action_name: "make_purchase_receipt",
3. in the form view. click the relevant custom button, the code defined in custom server action is called!

4. Attached is the screenshots for buttons in purchase order to Make-> Receipt,Make-> Invoice, Add Items From->Material Request
![purchase_order_js](https://user-images.githubusercontent.com/12823863/35960972-cd0346f2-0ce6-11e8-87c4-eb04c62c12ba.png)
![purchase order make receipt](https://user-images.githubusercontent.com/12823863/35960973-cd59f95c-0ce6-11e8-8eb9-9588994d5d1f.png)
![custom server action make stock entry](https://user-images.githubusercontent.com/12823863/35960974-cd9908fe-0ce6-11e8-9ba4-a4e4f6f8dfd9.png)
![custom server action material request](https://user-images.githubusercontent.com/12823863/35960975-cdd3b828-0ce6-11e8-8d95-02e4a23a81bf.png)

Commit:
Update __init__.py

Commit:
Update custom_server_action.py

szufisher commented:
Here documentation of creating, updating documents and executing python code when events triggerred
1. define new custom server action with following key information, it is possible to define multi custom server action per document type
   Action Type: Update Record
                        Create New Record
                        Execute Python Code
  Document Type:  specify the the document type from which the event will be triggerred
  Trigger On: the event to be triggerred
  Condition: the python expression to be evaluated, the current doc can be referenced, see condition example 
![image](https://user-images.githubusercontent.com/12823863/35977534-584bc0c2-0d1e-11e8-8581-2b4bfd5933b6.png)
  Target Document Type: default as the document type, the document type (options) of the link field can be specified, if target document type selected, the link field(which already filtered by target document) should be selected also
  Link field: only available for input when target document type is not same as document type
  Value Mapping: this is the child table with following fields
         Field: the field from the target document
         Type: either Value or Formula
         Value: when type is Value, it is treated as pure text, when type is Formula, the text in value field will be treated as python expression just like the condition field as above

   Code: the python code to be executed when Action Type is Execute Python Code, the following context variable is available: 
          doc : the current document( document type)
          update_doc: the document of the target document type to be updated
          frappe
          _
         json
        msgprint: the equivalent of frappe.throw, use msgprint to throw validation error instead, otherwise the exception raised by frappe.throw can not be caught and handled by app.application(request), seems the dynamically evaluated python code is not in the same call stack?

p.s except the Value Mapping, target document type, link field and Code, the other fields are the exact copy from email alerts, for more details, you can refer to email alerts

2. do whatever operations:SAVE/Submit/Cancel as specified in the Trigger On field on the document of document type, check whether the actions defined in the custom server action actually executed as expected,
    
3. refer to following sample test cases to have an idea how it really works

![custom_server_action_validation](https://user-images.githubusercontent.com/12823863/35978624-37e5cea6-0d21-11e8-8449-aa88b7e57240.png)

![custom_server_action_validation failure](https://user-images.githubusercontent.com/12823863/35978623-37a69bf0-0d21-11e8-8500-a963b5f50ac7.png)


![custom_server_action_update other doc](https://user-images.githubusercontent.com/12823863/35978622-37677146-0d21-11e8-9fd9-7445d23e5c1c.png)


![new lead existing customer](https://user-images.githubusercontent.com/12823863/35978628-389c0f86-0d21-11e8-9f2b-2d14b2253581.PNG)
![new lead existing customer linked back to lead](https://user-images.githubusercontent.com/12823863/35978627-385a6464-0d21-11e8-9230-2e57f8c67e75.PNG)

szufisher commented:
@nabinhait can you help to review this pull request?

rmehta commented:
@szufisher I would still love to see tests before merging this!

Also you need to update docs in the `/docs` folder.

General comment: What is your motivation for this function? We are doing SAAS for so many years without this.

Ultimately adding more and more custom action, makes it easier to achieve goals, yes, but it also means less motivation for contributions. Can you share some use cases where this can be used?

szufisher commented:
@rmehta 
1. what is the exact test required? the test py file? or the documentation of the test steps with screen shot and explanation in system?
2. /docs can be updated
3. the motivation is to provide useful and reliable tools to service providers and professional key users for more deep customization even on SAAS, this feature is already available on other popular systems like ODOO, SAP and ZOHO, it is more or less a advertised advanced features. 
4. Experience shows no such feature really needed does not mean it is not useful or needed in the future, but unfortunately so far I have not yet get any real for ERPNext project.

rmehta commented:
@szufisher 

> what is the exact test required? the test py file? or the documentation of the test steps with screen shot and explanation in system?

Server side tests to make sure your code does with it is supposed to do, along with tests for ensuring malicious code does not get executed.

> /docs can be updated

👍 

> the motivation is to provide useful and reliable tools to service providers and professional key users for more deep customization even on SAAS, this feature is already available on other popular systems like ODOO, SAP and ZOHO, it is more or less a advertised advanced features. Experience shows no such feature really needed does not mean it is not useful or needed in the future, 

I agree it could be useful, but is it worth the risk? We have more than 600 SAAS customers and it has never been a deal breaker. ERPNext has good REST API so you can use client side scripting to run authenticated server code.

> but unfortunately so far I have not yet get any real for ERPNext project.

Lets keep it on hold till then?

jk2081 commented:
I think this could be very useful for our installation especially since we are trying to bring in data from different systems into ERPNext which requires python code execution based on events on the doctype. These include biometric attendance systems, our serial interface with the weighing system etc...

Olawale1 commented:
@szufisher this is really awesome work; you obviously put in a great deal of effort... two thumbs up!!!

paurosello commented:
Thumbs up for the work!

I see the value of this for users of cloud providers, but this opens a big
door for vulnerabilities and security threats that I would be wary to
enable if I was a Cloud Provider.
Some kind of verified apps could be a better approach but this introduces
other problems and it´s not the road that Frappe wants to go.

Difficult approach unfortunately...


2018-04-05 12:26 GMT+02:00 Olawale1 <notifications@github.com>:

> @szufisher <https://github.com/szufisher> this is really awesome work;
> you obviously put in a great deal of effort... two thumbs up!!!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/frappe/frappe/pull/4941#issuecomment-378890314>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABFNgt0sNw5sW5wF7uf2r0oRokUTcp36ks5tlfE7gaJpZM4RyZb1>
> .
>

paurosello commented:
Maybe http://restrictedpython.readthedocs.io/en/latest/ might help trying to sandbox python code to limited modules

rmehta commented:
I am closing this for inacivity. I think this is a great example for someone to pick up incase this requirement comes up again.

MaxMorais commented:
@szufisher @rmehta, I'm here because I'm working into an similar implementation, but more secure.

I'lve been working with dukpy (https://github.com/amol-/dukpy) into the last 18 months for handle complex data transformation from multile json sources using jsonata (http://jsonata.org/)

Last days, I needed one real scripting engine into the backend of frappe, for I handle some special cases.

So I build this:

https://www.useloom.com/share/da23a0d1eaab4c27914ca77b405f2407

I'm refining some little features, but I think, with that code, same approach can be achived, but in a more secure way, to support also, users into the frappe-cloud.

I defined, an very controlled way, and refined way to define which functions are available trought the scripting engine, using 2 new hooks.

![image](https://user-images.githubusercontent.com/1288676/43818665-026840a6-9ab6-11e8-93e7-991bf764b57b.png)

`studio_functions` enumerate all functions from python, that we are enabling into the JavaScript engine, so, as that way, we can have an real access control to the resources

`studio_library_path` allow the javascript engine to load any javascript module into the defined path, using only `require("my_lib")`;

The last thing, I'll finish, is the syncronization of the hooks and the javascript engine, to allow the users, to send they validations to the backend also.

Still you interested on that feature?

Also, this feature runs only with 200 LOCS

MaxMorais commented:
@rmehta of course I'll, I'm just finishing little tests, and I'll open an pull request into frappe!

Ofcourse, I'll need some guidance to build tests ;)

MaxMorais commented:
@rmehta, the app is here https://github.com/mxmo-co/jstudio, for now, it's on a primary staging!
There's some things that I'm working on, like event subscription (when doc.run_method is called)
Schedulling of the functions.

The functions and modules from frappe that are supported are listed here https://github.com/mxmo-co/jstudio/blob/master/studio/hooks.py#L136.

Also, I'm working in an way to get an minimal documentation based on function signature.

Any comment will be apreciated.
Also, when I finish the test phase, I'll prepare the full pull request.

rmehta commented:
@MaxMorais looks cool! Please add some docs to your README so I can evaluate what you are doing.
